REPL: Improve :kind by only accepting type expressions, Fixes SI-8529:

eed3si9n opened the pull request:
Fixes scala/bug#8529

The :kind implementation added in scala/scala#2340 was buggy and misleading in part because we could not find a robust way to turn a type expression string into a `Type`. See also [Getting type information inside scala repl via IMain](http://stackoverflow.com/questions/15678616/getting-type-information-inside-scala-repl-via-imain/15694003#15694003). The problem was that we didn't know how to pick up types defined or imported in the REPL, so #2340 faked it by treating the given type expression as a term.

That's the major source of problems reported in scala/bug#8529. This PR improves the String-to-type-symbol conversion by using `exprTyper.typeOfTypeString` repeatedly with `[Nothing]`. It's hacky, but a whole class of "expression" can now be called out as not being a type.

### rejection of terms

Before:

```scala
scala> :kind 5
scala.Int's kind is A
```

After:

```scala
scala> :kind 5
<console>: error: type 5 was not found
```

### fixes companion object mixup

Before:

```scala
scala> :kind -v Predef.Pair
scala.Predef.Pair's kind is A
*
This is a proper type.
```

After:

```scala
scala> :kind -v Predef.Pair
Pair's kind is F[+A1,+A2]
* -(+)-> * -(+)-> *
This is a type constructor: a 1st-order-kinded type.
```

### fixes type application not being treated as a proper type

Before:

```scala
scala> :kind -v Predef.Pair[Int, Int]
scala.Tuple2's kind is F[+A1,+A2]
* -(+)-> * -(+)-> *
This is a type constructor: a 1st-order-kinded type.
```

After:

```scala
scala> :kind -v Predef.Pair[Int, Int]
Pair[Int,Int]'s kind is A
*
This is a proper type.
```

### fixes partially applied type alias

Before:

```scala
scala> type IntTuple[+A] = (Int, A)
defined type alias IntTuple

scala> :kind IntTuple
<console>:12: error: not found: value IntTuple
       IntTuple
       ^
```

After:

```scala
scala> type IntTuple[+A] = (Int, A)
defined type alias IntTuple

scala> :kind IntTuple
IntTuple's kind is F[+A]
```

### fixes type lambdas

Before:

```scala
scala> :kind ({type l[A] = Either[Int, A]})#l
<console>:1: error: ';' expected but '#' found.
({type l[A] = Either[Int, A]})#l
                              ^

scala> trait Nat[-F1[_], +F2[_]] {}
defined trait Nat

scala> type ~>[-F1[_], +F2[_]] = Nat[F1, F2]
defined type alias $tilde$greater

scala> :kind ({type l[F[-_]] = F ~> List})#l
<console>:1: error: ';' expected but '#' found.
({type l[F[-_]] = F ~> List})#l
                             ^
```

After:

```scala
scala> :kind ({type l[A] = Either[Int, A]})#l
scala.util.Either[Int,?]'s kind is F[+A]

scala> trait Nat[-F1[_], +F2[_]] {}
defined trait Nat

scala> type ~>[-F1[_], +F2[_]] = Nat[F1, F2]
defined type alias $tilde$greater

scala> :kind ({type l[F[-_]] = F ~> List})#l
Nat[?,[+A]List[A]]'s kind is X[-F[A]]
```

### picking things up from REPL

```scala
scala> trait Functor[F[_]] {}
...
defined trait Functor

scala> :kind -v Functor
Functor's kind is X[F[A]]
(* -> *) -> *
This is a type constructor that takes type constructor(s): a higher-kinded type.

scala> object Bar { class Bop }
defined object Bar

scala> import Bar.Bop
import Bar.Bop

scala> :kind Bop
Bar.Bop's kind is A
```

### helpful help message

I added a bit of documentation in `:help` as well.

```
scala> :help kind

:kind [-v] <type>
Displays the kind of a given type.

    -v      Displays verbose info.

Kind is a system to describe types and type constructors, which are themselves also types.
`Int`, `Option[Int]` and every other types that you can make a value out of are called
a "proper type" and denoted as `A` using Scala notion, or as * symbol.

    scala> :kind Option[Int]
    Option[Int]'s kind is A

In the above `Option` is an example of a first-order type constructor, which is denoted as
`F[A]` using Scala notation, or * -> * using the star notation. Because of variance,
it's actually `F[+A]`.

    scala> :k -v Option
    Option's kind is F[+A]
    * -(+)-> *
    This is a type constructor: a 1st-order-kinded type.

When you have more complicated type parameters, :kind command can be used to find out
what you need to pass in.

    scala> trait ~>[-F1[_], +F2[_]] {}
    scala> :kind ~>
    ~>'s kind is X[-F1[A1],+F2[A2]]

This shows that `~>` accepts something of `F[A]` kind such as `List` and `Vector`.
```

eed3si9n committed:
Improve :kind by only accepting type expressions

Fixes scala/bug#8529

The :kind implementation added in scala/scala#2340 was buggy and misleading in part because we could not find a robust way to turn a type expression string into a Type. See also [Getting type information inside scala repl via IMain](http://stackoverflow.com/questions/15678616/getting-type-information-inside-scala-repl-via-imain/15694003#15694003). The problem was that we didn't know how to pick up types defined or imported in the REPL, so #2340 faked it by treating the given type expression as a term.

That's the major source of problems reported in scala/bug#8529. This commit improves the String to type symbol conversion by using `exprTyper.typeOfTypeString` repeatedly with `[Nothing]`. It's hacky, but a whole class of "expression" can now be called out as not being a type.

Before:

    scala> :kind 5
    scala.Int's kind is A

After:

    scala> :kind 5
    <console>:1: error: identifier expected but integer literal found.
    def $ires23: 5 = ???
                 ^

More importantly, it will no longer mix up companion objects with actual types.

Before:

    scala> :kind -v Predef.Pair
    scala.Predef.Pair's kind is A
    *
    This is a proper type.

After:

    scala> :kind -v Predef.Pair
    scala.Tuple2's kind is F[+A1,+A2]
    * -(+)-> * -(+)-> *
    This is a type constructor: a 1st-order-kinded type.

eed3si9n committed:
Implement :kind handling for type lambdas

When a type lambda is applied with concrete argument, for eample:

    def x: ({type l[A] = Either[Int, A]})#l[Nothing] = ???
    x: scala.util.Either[Int,Nothing]

The return type becomes an applied type with `Int` and `Nothing` supplied, and I was not able to retrieve the type lambda.
To workaround this issue, I am supplying `WildcardType` to represent the hole, calculate the kind as if no type parameters are supplied, and adjust the kind afterwards.

eed3si9n committed:
Update REPL's :help text for :kind

Previous implementation of :kind incorrectly accepted term expressions.
This is now fixed, so the help message is updated to reflect that.
I've also added more detailed text under `:help kind`.

Jasper-M commented:
Is it possible to suppress output like this?
```
<console>:1: error: identifier expected but integer literal found.
def $foo: 5 = ???
          ^
```
I think the REPL should try to expose as little of its internals as possible to the user.

adriaanm commented:
> Is it possible to suppress output like this?

Yes, see beQuietDuring/beSilentDuring/printResults.

eed3si9n commented:
> Is it possible to suppress output like this?

Most internal interpreter calls are in fact suppressed. I'm displaying the error message in question intentionally when a type was not found like this (note `false` that's passed in):

```diff
+        case NoType =>
+          exprTyper.typeOfExpression(s"def $$foo: $typeString = ???", false)
+          echo(":kind requires a type.")
```

This is analogous to `:type` command letting know the user that the term expression is not valid:

```scala
scala> :type "1" / 0
<console>:12: error: value / is not a member of String
       "1" / 0
           ^

scala> :type Foo
<console>:12: error: not found: value Foo
       Foo
       ^
```

I'd be happy to change to something like

```scala
scala> :kind 5
<console>:1: error: not found type 5
```

if you guys think that's better.

eed3si9n committed:
Clean up :kind's error message per review

Before:

    <console>:1: error: identifier expected but integer literal found.
    def $foo: 5 = ???
              ^
    :kind requires a type.

After:

    <console>: error: type 5 was not found

eed3si9n commented:
btw I've added d465df421e1caebcca89db577933531c49ee8c6e last week so the error message is simpler:

```
scala> :kind 5
<console>: error: type 5 was not found
```

Jasper-M commented:
FWIW I think it's better this way.

SethTisue commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -295,14 +295,47 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+       |    scala> :kind ~>
+       |    ~>'s kind is X[-F1[A1],+F2[A2]]
+       |
+       |This shows that `~>` accepts something of `F[A]` kind such as `List` and `Vector`.""".stripMargin
```
suggested revised text:

```text
Displays the kind of a given type.

    -v      Displays verbose info.

"Kind" is a word used to classify types and type constructors
according to their level of abstractness.

Concrete, fully specified types such as `Int` and `Option[Int]`
are called "proper types" and denoted as `A` using Scala
notation, or with the `*` symbol.

    scala> :kind Option[Int]
    Option[Int]'s kind is A

In the above, `Option` is an example of a first-order type
constructor, which is denoted as `F[A]` using Scala notation, or
* -> * using the star notation. `:kind` also includes variance
information in its output, so if we ask for the kind of `Option`,
we actually see `F[+A]`:

    scala> :k -v Option
    Option's kind is F[+A]
    * -(+)-> *
    This is a type constructor: a 1st-order-kinded type.

When you have more complicated types, `:kind` can be used to find
out what you need to pass in.

    scala> trait ~>[-F1[_], +F2[_]] {}
    scala> :kind ~>
    ~>'s kind is X[-F1[A1],+F2[A2]]

This shows that `~>` accepts something of `F[A]` kind, such as
`List` or `Vector`.
```

eed3si9n replied:
Thanks for the revision. I'll adopt it.

retronym replied:
I've pushed that change to your branch.

eed3si9n replied:
Awesome. Thanks for the merge!

retronym committed:
Improve help message for :kind in the REPL
